### PR TITLE
V2 Updates

### PR DESCRIPTION
**Pull request by kimdhamilton**
Bring up to [v2 verification standards](http://www.blockcerts.org/guide/verification-process.html), most significantly, adding authenticity checks to verify the issuer address from the transaction checks out with those on the issuer id page. We want to ensure that the tx was issued when the issuing key was valid.

Exceptional cases handled:

- tx signing key not on issuer id page
- tx signing key was revoked, expired, or not yet created 

Added more test cases to: cert-walletTests/CertificateValidationRequestTests.swift

- tampered v2 certificate: changed existing field should break hash and fail
- authenticity failure: certificate issued after issuer revoked key should fail
- valid v2 certificate with legacy (v1) issuer format should pass

There is a lot of code duplication in that test class, but I'm having issues getting the test expectations cleanly refactored. I did some cleanup to make this easier in the future.

Opened issue #32 to track detection of unmapped JSON-LD fields. This will be doable; just need to find the right way to pass the fallback context to swift ld libraries during normalization.

Note I changed assertion id (uid) to extract just the guid to allow for the following cases:
- it's a url, like we had in v1
- it has a `urn:uuid` prefix, which v2 allows (but also allows urls). The prefix is used for unhosted assertions.

### CODE REVIEW

File: Frameworks/BlockchainCertificates/CertificateV2.swift
Comment (bedinotti) [inline Frameworks/BlockchainCertificates/CertificateV2.swift:95]: This doesn't look right. Do v2 certificates not have signatures?

File: Frameworks/BlockchainCertificates/Issuer.swift
Comment (bedinotti) [inline Frameworks/BlockchainCertificates/Issuer.swift]: Just a Swift tip here. Since you've specified the type of the enum, you can leave off the enum type in the assignment

```
asVersion version: IssuerVersion = .one) throws {
```

File: Frameworks/BlockchainCertificates/Issuer.swift
Comment (bedinotti) [inline Frameworks/BlockchainCertificates/Issuer.swift:22]: yay default parameters!

File: Frameworks/BlockchainCertificates/Issuer.swift
Comment (bedinotti) [inline Frameworks/BlockchainCertificates/Issuer.swift]: Do you mean to fix this before merging?

Comment (bedinotti) [reply Frameworks/BlockchainCertificates/Issuer.swift]: Oh, but to answer your question, yes, it's used. It's so we can use `NSCoding` to save a list of Issuers to disk & load them again. Structs can't conform to `NSCoding` (only `class`es can), but providing a means of converting to-and-from a dictionary gets around that (since dictionaries are classes that support `NSCoding`)

File: Frameworks/BlockchainCertificates/Issuer.swift
Comment (bedinotti) [inline Frameworks/BlockchainCertificates/Issuer.swift]: Is this the only line different between `parseKeysV1` and `parseKeysV2`? Maybe it could be passed in as an argument?

File: Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift
Comment (bedinotti) [inline Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift]: Do you mean to address this before merging?

File: Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift
Comment (bedinotti) [inline Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift]: Looks like the UID regexp is used in a few places. Might be good to just have this in our `Constants` file. (Mostly noting for myself, unless you feel ambitious)

File: Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift
Comment (bedinotti) [inline Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift]: More Swift style. `()`s aren't needed in an if statement.

File: cert-walletTests/Resources/IssuerKey.swift
Comment (bedinotti) [inline cert-walletTests/Resources/IssuerKey.swift]: This should be in the BlockchainCertificates framework, I believe.

Comment (bedinotti) [reply cert-walletTests/Resources/IssuerKey.swift]: Wait, actually, why is this file needed in the test resources?

**Review by bedinotti — APPROVED** [review]
Looks like you're getting way more comfortable writing Swift. Well done, Kim! :shipit:

File: Frameworks/BlockchainCertificates/CertificateV2.swift
Comment (kimdhamilton) [reply Frameworks/BlockchainCertificates/CertificateV2.swift:95]: that's right -- the signature is the one from the transaction!

File: Frameworks/BlockchainCertificates/Issuer.swift
Comment (kimdhamilton) [reply Frameworks/BlockchainCertificates/Issuer.swift]: thanks -- I'll delete the comment

File: Frameworks/BlockchainCertificates/Issuer.swift
Comment (kimdhamilton) [reply Frameworks/BlockchainCertificates/Issuer.swift]: this is a goof -- cleaning up

File: Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift
Comment (kimdhamilton) [reply Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift]: I opened an issue for that https://github.com/blockchain-certificates/cert-wallet/issues/32

comment removed

File: Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift
Comment (kimdhamilton) [reply Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift]: ooh, sorry! meant to ask a good place to put it. updated

File: Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift
Comment (kimdhamilton) [reply Frameworks/BlockchainCertificates/Requests/CertificateValidationRequest.swift]: updated here and similar other places

File: cert-walletTests/Resources/IssuerKey.swift
Comment (kimdhamilton) [reply cert-walletTests/Resources/IssuerKey.swift]: Bad add. I removed it

File: Frameworks/BlockchainCertificates/Issuer.swift
Comment (kimdhamilton) [reply Frameworks/BlockchainCertificates/Issuer.swift]: done

File: Frameworks/BlockchainCertificates/Issuer.swift
Comment (kimdhamilton) [reply Frameworks/BlockchainCertificates/Issuer.swift]: fixed to pass keyRotationSchedule/keyRotationScheduleV2 as a function argument to single (unified) parseKeys function

**Comment by kimdhamilton** [comment]
pushed fixes for PR feedback. Confirmed tests pass locally. Waiting for travis build, and will merge if everything looks good

**Pull request merged**
